libsplinter/src/admin/store/diesel/models.rs

```diff
@@ -0,0 +1,569 @@
+    pub circuit_id: String,
+    pub node_id: String,
+}
+
+impl From<&ProposedCircuit> for Vec<ProposedNodeModel> {
+    fn from(proposed_circuit: &ProposedCircuit) -> Self {
+        proposed_circuit
+            .members
+            .iter()
+            .map(|node| ProposedNodeModel {
+                circuit_id: proposed_circuit.circuit_id.clone(),
+                node_id: node.node_id.clone(),
+            })
+            .collect()
+    }
+}
+
+/// Database model representation of the endpoint values associated with a `ProposedNode`
+#[derive(Debug, PartialEq, Associations, Identifiable, Insertable, Queryable, QueryableByName)]
+#[table_name = "proposed_node_endpoints"]
+#[belongs_to(ProposedCircuitModel, foreign_key = "circuit_id")]
+#[belongs_to(ProposedNodeModel, foreign_key = "node_id")]
+#[primary_key(circuit_id, node_id, endpoint)]
+pub struct ProposedNodeEndpointsModel {
+    pub circuit_id: String,
+    pub node_id: String,
+    pub endpoint: String,
+}
+
+impl From<&ProposedCircuit> for Vec<ProposedNodeEndpointsModel> {
+    fn from(proposed_circuit: &ProposedCircuit) -> Self {
+        let mut endpoint_models = Vec::new();
+        for node in &proposed_circuit.members {
+            endpoint_models.extend(
+                node.endpoints
+                    .iter()
+                    .map(|endpoint| ProposedNodeEndpointsModel {
+                        circuit_id: proposed_circuit.circuit_id.clone(),
+                        node_id: node.node_id.clone(),
+                        endpoint: endpoint.clone(),
+                    })
+                    .collect::<Vec<ProposedNodeEndpointsModel>>(),
+            );
+        }
+        endpoint_models
+    }
+}
+
+/// Database model representation of a `ProposedService`
+#[derive(Debug, PartialEq, Associations, Identifiable, Insertable, Queryable, QueryableByName)]
+#[table_name = "proposed_services"]
+#[belongs_to(ProposedCircuitModel, foreign_key = "circuit_id")]
+#[primary_key(circuit_id, service_id)]
+pub struct ProposedServiceModel {
+    pub circuit_id: String,
+    pub service_id: String,
+    pub service_type: String,
+}
+
+impl From<&ProposedCircuit> for Vec<ProposedServiceModel> {
+    fn from(proposed_circuit: &ProposedCircuit) -> Self {
+        proposed_circuit
+            .roster
+            .iter()
+            .map(|service| ProposedServiceModel {
+                circuit_id: proposed_circuit.circuit_id.clone(),
+                service_id: service.service_id.clone(),
+                service_type: service.service_type.clone(),
+            })
+            .collect()
+    }
+}
+
+/// Database model representation of the `allowed_nodes` associated with a `ProposedService`
+#[derive(Debug, PartialEq, Associations, Identifiable, Insertable, Queryable, QueryableByName)]
+#[table_name = "proposed_services_allowed_nodes"]
+#[belongs_to(ProposedCircuitModel, foreign_key = "circuit_id")]
+#[belongs_to(ProposedServiceModel, foreign_key = "service_id")]
```

Comment:
Yeah, looking at the documentation composite foreign keys aren't supported which I took differently from what was being represented here, so this was more an attempt to see if it would compile / be helpful in the operations. So far, though, it's not. So, going to keep the most direct relationship (and remove the secondary one)